[api-minor] Improve performance with image masks (bug 857031)

calixteman opened the pull request:
- it aims to partially fix performance issue reported: https://bugzilla.mozilla.org/show_bug.cgi?id=857031;
- the idea is too avoid to use byte arrays but use ImageBitmap which are a way faster to draw:
  * an ImageBitmap is Transferable which means that it can be built in the worker instead of in the main thread:
    - this is achieved in using an OffscreenCanvas when it's available, there is a bug to enable them
      for pdf.js: https://bugzilla.mozilla.org/show_bug.cgi?id=1763330;
    - or in using createImageBitmap: in Firefox a task is sent to the main thread to build the bitmap so
      it's slightly slower than using an OffscreenCanvas.
  * it's transfered from the worker to the main thread by "reference";
  * the byte buffers used to create the image data have a very short lifetime and ergo the memory used is globally
    less than before.
- Use the localImageCache for the mask;
- Fix the pdf issue4436r.pdf: it was expected to have a binary stream for the image;
- Move the singlePixel trick from operator_list to image: this way we can use this trick even if it isn't in a set
  as defined in operator_list.

calixteman committed:
Improve performance with image masks (bug 857031)
- it aims to partially fix performance issue reported: https://bugzilla.mozilla.org/show_bug.cgi?id=857031;
- the idea is too avoid to use byte arrays but use ImageBitmap which are a way faster to draw:
  * an ImageBitmap is Transferable which means that it can be built in the worker instead of in the main thread:
    - this is achieved in using an OffscreenCanvas when it's available, there is a bug to enable them
      for pdf.js: https://bugzilla.mozilla.org/show_bug.cgi?id=1763330;
    - or in using createImageBitmap: in Firefox a task is sent to the main thread to build the bitmap so
      it's slightly slower than using an OffscreenCanvas.
  * it's transfered from the worker to the main thread by "reference";
  * the byte buffers used to create the image data have a very short lifetime and ergo the memory used is globally
    less than before.
- Use the localImageCache for the mask;
- Fix the pdf issue4436r.pdf: it was expected to have a binary stream for the image;
- Move the singlePixel trick from operator_list to image: this way we can use this trick even if it isn't in a set
  as defined in operator_list.

calixteman commented:
For information, I tested with the pdf in https://bugzilla.mozilla.org/show_bug.cgi?id=878397.
I measured the memory used with the devtools in local Firefox build:
 - current: 24.44Mb when the pdf is just loaded and 18.50Mb after a GC
 - with the patch and OffscreenCanvas enabled: 10.07Mb after pdf loaded an 7.49Mb after a GC.

Snuffleupagus reviewed (changes requested):
Overall I think this looks like a really nice way to improve things here, however there's unfortunately two problematic areas (which is why I holding off on doing a "proper" review).

For both Safari and Node.js, none of the functionality that this patch depends on is actually supported :-(
 - https://developer.mozilla.org/en-US/docs/Web/API/OffscreenCanvas#browser_compatibility, which is not available in either environment.
 - https://developer.mozilla.org/en-US/docs/Web/API/createImageBitmap#browser_compatibility, only available in Safari 15 but *not* in Workers.
 - https://developer.mozilla.org/en-US/docs/Web/API/ImageData/ImageData#browser_compatibility, which is not available in Node.js environments.

Hence it, most unfortunately, looks to me like we'll actually need to keep the old code around as a fallback and [do something similar to this existing code](https://github.com/mozilla/pdf.js/blob/27e738dff951160420575216c080e22027af0a86/src/display/canvas.js#L648-L652) in the `putBinaryImageMask` function as well.

calixteman commented:
Arf...
A way to workaround this without the need to maintain too much stuff could be to generate a PNG buffer in using (this is based on a @brendandahl's idea):
https://github.com/mozilla/pdf.js/blob/27e738dff951160420575216c080e22027af0a86/src/display/svg.js#L81
then convert into a blob and pass it to `sendImgData` where we can easily create an `ImageBitmap`.

Before using `createImageBitmap` I was sending to the main thread a buffer (consumed "immediately" in https://github.com/mozilla/pdf.js/blob/27e738dff951160420575216c080e22027af0a86/src/display/api.js#L2764 in using `createImageBitmap`) in case of no `OffscreenCanvas` and with the pdf from bug [878397](https://bugzilla.mozilla.org/show_bug.cgi?id=878397) I noticed that before GC we have a huge memory consumption (more than 100Mb).
Maybe I won't have the same issue with a `Blob` else, as you mentioned, there still is this possibility to keep the current way to do.

marco-c commented:
Is this also improving perf with the PDF from https://bugzilla.mozilla.org/show_bug.cgi?id=1135277?

Snuffleupagus commented on src/core/evaluator.js:
```diff
@@ -612,16 +615,75 @@ class PartialEvaluator {
+        this._sendImgData(objId, imgData);
+      }
+      operatorList.addDependency(objId);
+      args = [{ data: objId, width: w, height: h }];
```
You're not passing along the `interpolate` parameter here.

While it might not be necessary for this particular data, it still seems nice to be consistent w.r.t. the API. Also, I'd probably suggest actually returning the parameters from the `PDFImage.createMask`-call itself to avoid any future "surprises" here:
```javascript
args = [{
  data: objId,
  width: imgData.width,
  height: imgData.height,
  interpolate: imgData.interpolate,
}];
```

Snuffleupagus commented on src/display/canvas.js:
```diff
@@ -1120,6 +1124,15 @@ class CanvasGraphics {
     this._cachedGetSinglePixelWidth = null;
   }
 
+  getObject(data, _default) {
```
How about `_default` -> `fallback` instead? 

Also, I'd suggest changing this to `getObject(data, fallback = null) {` since some of the call-sites below would then not need to provide a second argument any more.

Snuffleupagus commented on src/core/image.js:
```diff
@@ -288,7 +292,7 @@ class PDFImage {
   }
 
-  static createMask({
+  static createMaskForType3({
```
Given that this is no longer used only for Type3-fonts, could we perhaps use a slightly more generic name?
Naming things is always hard, but how about e.g. `createFullMask`, `createRawMask`, or some other (better) name?

Also, please remember to update the method name in the `assert` call used in this method.

Snuffleupagus reviewed (changes requested):
This looks very good in general, and keeping the old code as a fallback is a nice/simple solution for Node.js and older browsers.
However, there's a few things that I'd like to see changed, please see the inline comments.

Given that `OffscreenCanvas` is not enabled by default in Firefox, we're thus not testing that configuration as-is. That seems unfortunate, but I believe that we should be able to work-around this by adding `gfx.offscreencanvas.enabled: true,` to the list in https://github.com/mozilla/pdf.js/blob/497c0610cd0ab465965e096adbad46246983883e/test/test.js#L943-L963

Snuffleupagus commented on src/core/evaluator.js:
```diff
@@ -612,16 +615,75 @@ class PartialEvaluator {
+      } else {
+        this._sendImgData(objId, imgData);
+      }
+      operatorList.addDependency(objId);
```
Also, let's place this line *directly* after the `objId` definition above.

calixteman commented on src/core/image.js:
```diff
@@ -343,6 +347,80 @@ class PDFImage {
+        height,
+        inverseDecode,
+      });
+      bitmapPromise = createImageBitmap(imgData);
```
I made it async but in the general case it's sync and in a very particular case it's async: I thought it was useless add any potential overhead just for a corner case.
Since now we're falling back on the old way to do I wonder if it's worth having the createBitmapImage case, wdyt ?

Snuffleupagus replied:
> Since now we're falling back on the old way to do I wonder if it's worth having the createBitmapImage case, wdyt ?

That's an excellent point, let's just remove the `createBitmapImage`-case and keep this as a synchronous method then!

Snuffleupagus commented on src/core/evaluator.js:
```diff
@@ -620,8 +647,38 @@ class PartialEvaluator {
+        return;
+      }
+
       imgData.cached = !!cacheKey;
```
Does this still matter here, since the actual imgData will be sent a *separate* message (via `this._sendImgData`) and below `args.data` will always contain a string now?

Snuffleupagus commented on src/core/operator_list.js:
```diff
@@ -701,11 +670,12 @@ class OperatorList {
             );
           }
-          if (!arg.cached) {
+          if (!arg.cached && arg.data && arg.data.buffer) {
```
Nit: Maybe `if (!arg.cached && arg.data && arg.data.buffer instanceof ArrayBuffer) {` to prevent *any possible* future issues (since `arg.data` will be a string in some cases)?

Snuffleupagus reviewed (approved):
r=me, with a couple of last comments and all tests passing; thank you!

Snuffleupagus commented on src/core/image.js:
```diff
@@ -343,6 +351,69 @@ class PDFImage {
+
+      ctx.putImageData(imgData, 0, 0);
+      const bitmap = canvas.transferToImageBitmap();
+
```
At this point, I'm guessing that the OffscreenCanvas is no longer needed for anything once the `ImageBitmap` has been created?
If so, would it make sense to try and "dispose" of it *immediately* similar to what we do in other parts of the code-base; see e.g. https://github.com/mozilla/pdf.js/blob/2b673a694114268bad7cc7dd687bf004513e1815/web/pdf_page_view.js#L297-L300

calixteman replied:
I'm not sure it's useful, because we transfered the ownership of the underlying surface (in the ImageBitmap) associated with this canvas and the object itself should be destroyed at the end of the scope.
In following what happens when e.g. `SetWidth` is called:
https://searchfox.org/mozilla-central/rev/c0a97899404931861a4bbec21f55d3cd79b90bb8/dom/canvas/OffscreenCanvas.cpp#76
I don't see anything special which could indicate that's useful.
Anyway, I trust your judgement so if you think it's better, I'll add these lines.

Snuffleupagus replied:
I really have no idea if this matters for OffscreenCanvas, I was simply asking based on a pattern we use elsewhere for "regular" canvases (originally added all the way back in PR #4920).

So, do whatever you see fit here :-)

calixteman replied:
@jrmuizel, do you know if setting width & height to 0 is useful for an `OffscreenCanvas` with a lifetime limited to a function scope ?

calixteman commented:
/botio test

calixteman commented:
The failure in integration tests is unrelated to this patch.
And the few "regressions" are from my pov an improvement and are related to the change about 1x1 opaque masks.

Snuffleupagus commented:
> And the few "regressions" are from my pov an improvement and are related to the change about 1x1 opaque masks.
> @Snuffleupagus does it work for you ?

Generally yes, but why was modifying the `issue4436r.pdf` test-file necessary here? And, importantly, does the "new" file still trigger the *same exact* code-paths as before?

Note that the reduced test-case was originally created based on the information in https://github.com/mozilla/pdf.js/issues/4436#issuecomment-37643401

calixteman commented:
The stream for the image contains the 4 chars 0x00 and since the image is 1x1 then only the char "0" is taken which correspond to 48.
The 1x1 stuff I moved was executed only if the element was 0 for a 1x1 image so it wasn't executed.
Now only the first bit is considered, so the trick is correctly executed.
So I guessed that 0x00 was for a null byte, hence I replaced the 4 bytes by NULL (to avoid to have to rewrite the ref table). 
So as far as I can tell, the previous test case was not triggering the code path it was supposed to test.
It'll be ok to let it like it is, I mean with the 0x00, because the first bit is 0 of 48 and we'll have the same "regression", but I thought that it was better to have a test with the correct value.

Snuffleupagus commented:
> The stream for the image contains the 4 chars 0x00 and since the image is 1x1 then only the char "0" is taken which correspond to 48. The 1x1 stuff I moved was executed only if the element was 0 for a 1x1 image so it wasn't executed. Now only the first bit is considered, so the trick is correctly executed. So I guessed that 0x00 was for a null byte, hence I replaced the 4 bytes by NULL (to avoid to have to rewrite the ref table). So as far as I can tell, the previous test case was not triggering the code path it was supposed to test. It'll be ok to let it like it is, I mean with the 0x00, because the first bit is 0 of 48 and we'll have the same "regression", but I thought that it was better to have a test with the correct value.

Thanks for the thorough explanation, this seems like a good improvement here!

(I suppose this just shows that trying to create a reduced test-case without access to the original document is rarely a good idea...)

calixteman commented:
/botio makeref
